compileJava error on JDK 17 due to `java.lang.NoClassDefFoundError: com/sun/tools/javac/util/Filter`

## What happened?

<!--
Describe what you've observed and why it's bad (please include whatever stacktraces/version numbers you can).
Clear steps to reproduce the behaviour are always helpful too!
-->

In the following project on JDK 17 I get `java.lang.NoClassDefFoundError: com/sun/tools/javac/util/Filter` when I `gradlew clean build`: https://github.com/haisi/jPBE

```
 Task :compileJava
/Users/hasankara/projects/private/jPBE/src/main/java/li/selman/jpbe/datastructure/Graph.java:144: warning: [UnusedMethod] Private method 'getAllTraceExpressions' is never used.
    private List<Expressions> getAllTraceExpressions(Stream<Edge> edgeStream) {
                              ^
    (see https://errorprone.info/bugpattern/UnusedMethod)
  Did you mean to remove this line?
/Users/hasankara/projects/private/jPBE/src/main/java/li/selman/jpbe/datastructure/Graph.java:32: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
public class Graph {
       ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:
  
     error-prone version: 2.9.0
     BugPattern: ClassInitializationDeadlock
     Stack Trace:
     java.lang.NoClassDefFoundError: com/sun/tools/javac/util/Filter
        at com.palantir.baseline.errorprone.ClassInitializationDeadlock.matchClass(ClassInitializationDeadlock.java:71)
        at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:450)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:548)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:151)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:860)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:119)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:152)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:561)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:151)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:614)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:60)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:58)
        at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
        at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
        at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1394)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1341)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:933)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.invocationHelper(JavacTaskImpl.java:152)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
        at org.gradle.internal.compiler.java.IncrementalCompileTask.call(IncrementalCompileTask.java:89)
        at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:94)
        at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:54)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:39)
        at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
        at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
        at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:32)
        at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:22)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:85)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:55)
        at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
        at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:61)
        at java.base/java.lang.Thread.run(Thread.java:833)
  Caused by: java.lang.ClassNotFoundException: com.sun.tools.javac.util.Filter
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:440)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
        ... 59 more
1 error
1 warning

> Task :compileJava FAILED
```

On the [error prone side](https://github.com/google/error-prone/commit/5009d026a45ffacb2eb2ff82808fb5f076bf59a4), this JDK17 incompatibility has already been fixed.

<details closed>
<summary>A compatibility layer might look like:</summary>

```java
import com.sun.tools.javac.code.Scope;
import com.sun.tools.javac.code.Scope.LookupKind;
import com.sun.tools.javac.code.Symbol;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.util.Arrays;
import java.util.function.Predicate;
import javax.lang.model.element.Name;

/**
 * A compatibility wrapper around {@link com.sun.tools.javac.util.Filter}.
 * <p>
 * Adapted from <a href="https://github.com/google/error-prone/blob/master/check_api/src/main/java/com/google/errorprone/util/ErrorProneScope.java">
 *     Error Prone</a>.
 */
class ScopeCompat {
    private final Scope scope;

    ScopeCompat(Scope scope) {
        this.scope = scope;
    }

    @SuppressWarnings("unchecked") // reflection
    public Iterable<Symbol> getSymbolsByName(Name name, Predicate<Symbol> predicate) {
        return (Iterable<Symbol>) invoke(getSymbolsByName, name, maybeAsFilter(predicate));
    }

    @SuppressWarnings("unchecked") // reflection
    public Iterable<Symbol> getSymbolsByName(Name name, Predicate<Symbol> predicate, LookupKind lookupKind) {
        return (Iterable<Symbol>) invoke(getSymbolsByNameLookupKind, name, maybeAsFilter(predicate), lookupKind);
    }

    @SuppressWarnings("unchecked") // reflection
    public Iterable<Symbol> getSymbols(Predicate<Symbol> predicate) {
        return (Iterable<Symbol>) invoke(getSymbols, maybeAsFilter(predicate));
    }

    @SuppressWarnings("unchecked") // reflection
    public Iterable<Symbol> getSymbols(LookupKind lookupKind) {
        return (Iterable<Symbol>) invoke(getSymbols, lookupKind);
    }

    @SuppressWarnings("unchecked") // reflection
    public Iterable<Symbol> getSymbols(Predicate<Symbol> predicate, LookupKind lookupKind) {
        return (Iterable<Symbol>) invoke(getSymbols, maybeAsFilter(predicate), lookupKind);
    }

    public boolean anyMatch(Predicate<Symbol> predicate) {
        return (boolean) invoke(anyMatch, maybeAsFilter(predicate));
    }

    private static final Class<?> FILTER_CLASS = getFilterClass();

    private static Class<?> getFilterClass() {
        try {
            return Class.forName("com.sun.tools.javac.util.Filter");
        } catch (ClassNotFoundException e) {
            return null;
        }
    }

    private static final Method anyMatch = getImpl("anyMatch", Predicate.class);

    private static final Method getSymbolsByName = getImpl("getSymbolsByName", Name.class, Predicate.class);

    private static final Method getSymbolsByNameLookupKind =
            getImpl("getSymbolsByName", Name.class, Predicate.class, LookupKind.class);

    private static final Method getSymbols = getImpl("getSymbols", Predicate.class);

    private static Method getImpl(String name, Class<?>... parameters) {
        return FILTER_CLASS != null
                ? getMethodOrDie(
                        Scope.class,
                        name,
                        Arrays.stream(parameters)
                                .map(p -> p.equals(Predicate.class) ? FILTER_CLASS : p)
                                .toArray(Class<?>[]::new))
                : getMethodOrDie(Scope.class, name, parameters);
    }

    private Object invoke(Method method, Object... args) {
        try {
            return method.invoke(scope, args);
        } catch (ReflectiveOperationException e) {
            throw new LinkageError(e.getMessage(), e);
        }
    }

    private Object maybeAsFilter(Predicate<Symbol> predicate) {
        if (FILTER_CLASS == null) {
            return predicate;
        }
        return Proxy.newProxyInstance(
                getClass().getClassLoader(), new Class<?>[] {FILTER_CLASS}, (proxy, method, args) -> {
                    checkState(method.getName().equals("accepts"));
                    return predicate.test((Symbol) args[0]);
                });
    }

    private static Method getMethodOrDie(Class<?> clazz, String name, Class<?>... parameters) {
        try {
            return clazz.getMethod(name, parameters);
        } catch (NoSuchMethodException e) {
            throw new LinkageError(e.getMessage(), e);
        }
    }
}
```
</details>

<details closed>
<summary>With changes in `ClassInitalizationDeadlock#cannotBeInitializedExternally` </summary>

```java
private boolean cannotBeInitializedExternally(ClassSymbol newClassSymbol) {
    if (newClassSymbol.isPrivate()) {
        return true;
    }

    WriteableScope scope = newClassSymbol.members();
    return !new ScopeCompat(scope)
            .getSymbols(CanBeExternallyInitializedFilter.INSTANCE, LookupKind.NON_RECURSIVE)
            .iterator()
            .hasNext();
}
```

</details>

```
java -version
openjdk version "17" 2021-09-14
OpenJDK Runtime Environment Temurin-17+35 (build 17+35)
OpenJDK 64-Bit Server VM Temurin-17+35 (build 17+35, mixed mode, sharing)
```

## What did you want to happen?

Support error-prone on JDK 17.
<!--
Suggest better behaviour
-->

--- a/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/BaselineErrorProneScope.java
+++ b/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/BaselineErrorProneScope.java
@@ -32,6 +32,9 @@ import java.util.function.Predicate;
  * A compatibility wrapper around {@link com.sun.tools.javac.util.Filter}.
  * Adapted from {@link com.google.errorprone.util.ErrorProneScope} with additional methods.
  *
+ * Original code from (Apache-2.0 License):
+ * https://github.com/google/error-prone/blob/ce87ca1c7bb005371837b82ffa69041dd8a356e5/check_api/src/main/java/com/google/errorprone/util/ErrorProneScope.java
+ *
  * TODO(fwindheuser): Delete after upstreaming missing methods into "ErrorProneScope".
  */
 @SuppressWarnings("ThrowError")

--- a/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/BaselineErrorProneScope.java
+++ b/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/BaselineErrorProneScope.java
@@ -1,0 +1,107 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.errorprone.util.RuntimeVersion;
+import com.sun.tools.javac.code.Scope;
+import com.sun.tools.javac.code.Scope.LookupKind;
+import com.sun.tools.javac.code.Symbol;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.Arrays;
+import java.util.function.Predicate;
+
+/**
+ * A compatibility wrapper around {@link com.sun.tools.javac.util.Filter}.
+ * Adapted from {@link com.google.errorprone.util.ErrorProneScope} with additional methods.
+ *
+ * TODO(fwindheuser): Delete after upstreaming missing methods into "ErrorProneScope".
+ */
+@SuppressWarnings("ThrowError")
+public final class BaselineErrorProneScope {
+
+    @SuppressWarnings("unchecked") // reflection
+    public Iterable<Symbol> getSymbols(Predicate<Symbol> predicate, LookupKind lookupKind) {
+        return (Iterable<Symbol>) invoke(getSymbols, maybeAsFilter(predicate), lookupKind);
+    }
+
+    private static final Class<?> FILTER_CLASS = getFilterClass();
+
+    private static Class<?> getFilterClass() {
+        if (RuntimeVersion.isAtLeast17()) {
+            return null;
+        }
+        try {
+            return Class.forName("com.sun.tools.javac.util.Filter");
+        } catch (ClassNotFoundException e) {
+            throw new LinkageError(e.getMessage(), e);
+        }
+    }
+
+    private static final Method getSymbols = getImpl("getSymbols", Predicate.class, LookupKind.class);
+
+    private static Method getImpl(String name, Class<?>... parameters) {
+        return FILTER_CLASS != null
+                ? getMethodOrDie(
+                        Scope.class,
+                        name,
+                        Arrays.stream(parameters)
+                                .map(p -> p.equals(Predicate.class) ? FILTER_CLASS : p)
+                                .toArray(Class<?>[]::new))
+                : getMethodOrDie(Scope.class, name, parameters);
+    }
+
+    private final Scope scope;
+
+    BaselineErrorProneScope(Scope scope) {
+        this.scope = scope;
+    }
+
+    private Object invoke(Method method, Object... args) {
+        try {
+            return method.invoke(scope, args);
+        } catch (ReflectiveOperationException e) {
+            throw new LinkageError(e.getMessage(), e);
+        }
+    }
+
+    @SuppressWarnings("ProxyNonConstantType")
+    private Object maybeAsFilter(Predicate<Symbol> predicate) {
+        if (FILTER_CLASS == null) {
+            return predicate;
+        }
+        return Proxy.newProxyInstance(
+                getClass().getClassLoader(), new Class<?>[] {FILTER_CLASS}, new InvocationHandler() {
+                    @Override
+                    public Object invoke(Object proxy, Method method, Object[] args) {
+                        checkState(method.getName().equals("accepts"));
+                        return predicate.test((Symbol) args[0]);
+                    }
+                });
+    }
+
+    private static Method getMethodOrDie(Class<?> clazz, String name, Class<?>... parameters) {
+        try {
+            return clazz.getMethod(name, parameters);
+        } catch (NoSuchMethodException e) {
+            throw new LinkageError(e.getMessage(), e);
+        }
+    }
+}

--- a/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ClassInitializationDeadlock.java
+++ b/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ClassInitializationDeadlock.java
@@ -41,7 +41,6 @@ import com.sun.tools.javac.code.Symbol.ClassSymbol;
 import com.sun.tools.javac.code.Symbol.MethodSymbol;
 import com.sun.tools.javac.code.Symbol.VarSymbol;
 import com.sun.tools.javac.code.Type;
-import com.sun.tools.javac.util.Filter;
 
 @AutoService(BugChecker.class)
 @BugPattern(
@@ -130,9 +129,8 @@ public final class ClassInitializationDeadlock extends BugChecker implements Bug
             if (newClassSymbol.isPrivate()) {
                 return true;
             }
-            return !newClassSymbol
-                    .members()
-                    .getSymbols(CanBeExternallyInitializedFilter.INSTANCE, LookupKind.NON_RECURSIVE)
+            return !new BaselineErrorProneScope(newClassSymbol.members())
+                    .getSymbols(ClassInitializationDeadlock::canBeExternallyInitialized, LookupKind.NON_RECURSIVE)
                     .iterator()
                     .hasNext();
         }
@@ -229,28 +227,23 @@ public final class ClassInitializationDeadlock extends BugChecker implements Bug
                 && !ASTHelpers.isSameType(maybeSubtype, baseType, state);
     }
 
-    private enum CanBeExternallyInitializedFilter implements Filter<Symbol> {
-        INSTANCE;
-
-        @Override
-        public boolean accepts(Symbol symbol) {
-            switch (symbol.getKind()) {
-                case FIELD:
-                    // fall through
-                case METHOD:
-                    if (symbol.isStatic() && !symbol.isPrivate()) {
-                        return true;
-                    }
-                    break;
-                case CONSTRUCTOR:
-                    if (!symbol.isPrivate()) {
-                        return true;
-                    }
-                    break;
-                default:
-                    // fall through
-            }
-            return false;
+    private static boolean canBeExternallyInitialized(Symbol symbol) {
+        switch (symbol.getKind()) {
+            case FIELD:
+                // fall through
+            case METHOD:
+                if (symbol.isStatic() && !symbol.isPrivate()) {
+                    return true;
+                }
+                break;
+            case CONSTRUCTOR:
+                if (!symbol.isPrivate()) {
+                    return true;
+                }
+                break;
+            default:
+                // fall through
         }
+        return false;
     }
 }

--- a/changelog/@unreleased/pr-1936.v2.yml
+++ b/changelog/@unreleased/pr-1936.v2.yml
@@ -1,0 +1,5 @@
+type: fix
+fix:
+  description: Fix jdk17 incompatibility of ClassInitializationDeadlock
+  links:
+  - https://github.com/palantir/gradle-baseline/pull/1936

